factoryWithTypeCheckers.js

```diff
@@ -314,6 +315,36 @@ module.exports = function(isValidElement, throwOnDirectAccess) {
     return createChainableTypeChecker(validate);
   }
 
+
+  function createElementWithTypeChecker(expectedType) {
+    var ACCEPTABLE_TYPES_OF_EXPECTED_TYPES = ['string', 'function'];
+
+    if (ACCEPTABLE_TYPES_OF_EXPECTED_TYPES.indexOf(typeof expectedType) === -1) {
+      process.env.NODE_ENV !== 'production' ? warning(false, 'Invalid argument supplied to ElementWithType, expected an Html tag name or a Component.') : void 0;
```

Comment:
this should not use `warning` here; I'm suggesting this have a literal `throw` statement.

Reply:
this would be different from the behaviors of other type checkers.

https://github.com/congwenma/prop-types/blob/add-componentOfType/factoryWithTypeCheckers.js#L277

Reply:
hm, that seems like a missed opportunity.

Reply:
wouldn't this be a useful pattern by `not making a big deal` about prop types in production?

Reply:
I'm not sure what you mean; invalid props are bugs. The point of not warning about them in production is performance; the exception I'm hoping for here would only apply to dev anyways, since the entire propType validator is compiled out anyways in production.

Reply:
sorry I thought you mean `replacing the entire line with a  throw`. 
I think what you are suggesting can make things easier and we'd be removing dependency on `fbjs/warning`, however I think theres some logic in `fbjs/warning` that helps debugging prop-types and handle some special cases, and this probably warrants a separate PR.